token/program-2022/src/extension/confidential_mint_burn/instruction.rs

```diff
@@ -0,0 +1,415 @@
+            .map_err(|_| TokenError::InvalidState)?,
+        &new_supply_ciphertext,
+        &new_supply_opening,
+        current_supply,
+    )
+    .map_err(|_| TokenError::ProofGeneration)?;
+    accounts.push(AccountMeta::new_readonly(sysvar::instructions::id(), false));
+
+    Ok(vec![
+        encode_instruction(
+            token_program_id,
+            accounts,
+            TokenInstruction::ConfidentialMintBurnExtension,
+            ConfidentialMintBurnInstruction::RotateSupplyElGamal,
+            &RotateSupplyElGamalData {
+                new_supply_elgamal_pubkey: Some(Into::<PodElGamalPubkey>::into(
+                    *new_supply_elgamal_keypair.pubkey(),
+                ))
+                .try_into()?,
+                proof_location: 1,
+            },
+        ),
+        ProofInstruction::VerifyCiphertextCiphertextEquality.encode_verify_proof(None, &proof_data),
+    ])
+}
+
+/// Create a `ConfidentialMint` instruction
+#[allow(clippy::too_many_arguments)]
+#[cfg(not(target_os = "solana"))]
+pub fn confidential_mint_with_split_proofs(
+    token_program_id: &Pubkey,
+    token_account: &Pubkey,
+    mint: &Pubkey,
+    supply_elgamal_pubkey: Option<ElGamalPubkey>,
+    authority: &Pubkey,
+    multisig_signers: &[&Pubkey],
+    context_accounts: &MintSplitContextStateAccounts,
+    new_decryptable_supply: AeCiphertext,
+) -> Result<Instruction, ProgramError> {
+    check_program_account(token_program_id)?;
+    let mut accounts = vec![AccountMeta::new(*token_account, false)];
+    // we only need write lock to adjust confidential suppy on
+    // mint if a value for supply_elgamal_pubkey has been set
+    if supply_elgamal_pubkey.is_some() {
+        accounts.push(AccountMeta::new(*mint, false));
+    } else {
+        accounts.push(AccountMeta::new_readonly(*mint, false));
+    }
+
+    accounts.push(AccountMeta::new_readonly(
+        *context_accounts.equality_proof,
+        false,
+    ));
+    accounts.push(AccountMeta::new_readonly(
+        *context_accounts.ciphertext_validity_proof,
+        false,
+    ));
+    accounts.push(AccountMeta::new_readonly(
+        *context_accounts.range_proof,
+        false,
+    ));
+
+    accounts.push(AccountMeta::new_readonly(
+        *authority,
+        multisig_signers.is_empty(),
+    ));
+    for multisig_signer in multisig_signers.iter() {
+        accounts.push(AccountMeta::new_readonly(**multisig_signer, true));
+    }
+
+    Ok(encode_instruction(
+        token_program_id,
+        accounts,
+        TokenInstruction::ConfidentialMintBurnExtension,
+        ConfidentialMintBurnInstruction::ConfidentialMint,
+        &MintInstructionData {
+            new_decryptable_supply: new_decryptable_supply.into(),
+            proof_instruction_offset: 0,
+        },
+    ))
+}
+
+/// Context state accounts used in confidential mint
+#[derive(Clone, Copy)]
+pub struct BurnSplitContextStateAccounts<'a> {
```

Comment:
I actually switched to this for the mint as well [recently](https://github.com/solana-labs/solana-program-library/pull/7319/commits/18aa58dc845bc63aeecaa991e74f61b4911cbd67#diff-2153c862cbbf86482ac09472f08bd243c469c02e024b715f756be64242982f53L3485-L3489) since I thought it'd look more concise 🤦‍♂️ will change everything to ProofLocation then

Reply:
Alright changed it to use ProofLocation, I added a helper function [here](https://github.com/solana-labs/solana-program-library/pull/7319/files#diff-8de1088937c0dfb9e089b7e97ddc2be262714c3618c2372b4cc56b853bf89b0dR138-R179) since most of the `ProofLocation -> Instruction` logic was pretty much copy-paste, ptal

Reply:
Okay yeah, looks good to me. Thanks!